### ISSUE TITLE
Shallow Routing

### ISSUE THREAD

**Issue by dajinchu** [vikejs/vike#1231]
### Description

Next.js has https://nextjs.org/docs/pages/building-your-application/routing/linking-and-navigating#shallow-routing

Would it be possible for Vike's `navigate` function to support something similar?

Use case: we are trying to hook up [UseQueryParam](https://www.npmjs.com/package/use-query-params) but can't have the whole pageContext get refetched when a single query param is changed. We use react-query, so we already are setup to just refetch what is needed.

**Comment by brillout** [comment]
What's your use case from a UI perspective (e.g. persisting search queries as URL query param)?

How urgent is it?

**Comment by dajinchu** [comment]
> What's your use case from a UI perspective (e.g. persisting search queries as URL query param)?
> 
> How urgent is it?

Yeah a filter UI. Pretty urgent, though I am fine implementing something in userland. Is it safe for me to just call `pushState`? 

Another requirement we have is that only the components that read the query param rerender, so I'm not sure Vike can even handle it at the framework level. Maybe vike-react could. I am inspired by how [Tanstack Router handles params](https://tanstack.com/router/v1/docs/guide/search-params)

**Comment by brillout** [comment]
> Pretty urgent

Okay, let's find a quick solution first and clean it later.

> filter UI

I see, makes sense.

So you basically want any URL change caused by the user changing settings in the UI filter to preserve data/state, correct?

But you still want `onRenderClient()` to be fired in order to re-render the whole page, correct?

> Is it safe for me to just call `pushState`

Yea, I think that's a possibility (especially given the fact that a `history.pushState()` call can be attached with persisted data e.g. `{ vike: { data: false } }`). But I'm leaning more towards `navigate('/bla', { data: false })` which AFAICT is equivalent while being simpler.

I'm also thinking a sensible default would be to skip the whole `pageContext` refetching when only the query param changes.

I think I'll be able to implement somthing quickly once I know more about your use case.

**Comment by dajinchu** [comment]
> So you basically want any URL change caused by the user changing settings in the UI filter to preserve data/state, correct?
> But you still want onRenderClient() to be fired in order to re-render the whole page, correct?

Yes and yes.

> history.pushState() call can be attached with persisted data e.g. { vike: { data: false } })

Before you make any changes to Vike, I'm thinking I could call `history.pushState()` and manage the re-render within react with useState. Do you see any issues with that? Could be a quick way to get this working w/o having to touch Vike

**Comment by brillout** [comment]
> I'm thinking I could call `history.pushState()` and manage the re-render within react with useState. Do you see any issues with that?

A user tried before but he ran into an issue. Let me find that discussion.

**Comment by brillout** [comment]
> A user tried before but [he ran into issues](https://github.com/vikejs/vike/issues/1169).

 Let me see if I can quickly fix that (I've an idea).

**Comment by brillout** [comment]
CC @thomasjm @snake-py (FYI @dajinchu is a high-tier sponsor which explains why I'm prioritizing this feature request).

**Comment by brillout** [comment]
> I'm also thinking a sensible default would be to skip the whole `pageContext` refetching when only the query param changes.

I'm hesitating because ~~some~~ most query params depend on data. That said, I can see most users to want to manage such data dependency within React (e.g. `useState()` and [Telefunc](https://telefunc.com/)) instead of using Vike's data fetching mechanism. Maybe it is a good default, but then it should be configurable.

**Comment by dajinchu** [comment]
Yeah it depends on your apps data fetching solution. For simple apps that are using pageContext for everything, refetching pageContext is probably what you want. For us, we have multiple data fetches on a single page, so we use react-query with prefetched data in `+onBeforeRender` that gets dehydrated into pageContext, and need to just refetch some keys.

TL;DR I think the default should be refetch and it can be turned off inside `navigate`. I don't think a global default switch is needed.

**Comment by dajinchu** [comment]
Re: `shouldVikeProcessPopState ` in the [discussion you linked.](https://github.com/vikejs/vike/issues/1169#issuecomment-1761331578)

Just FYI. We are already using a hack in `onBeforeRoute` to detect Vike's onpopstate in order to add custom behavior to back button navigation: https://github.com/Alignable/bifrost/blob/master/bifrost/renderer/onBeforeRoute.ts#L9

This is in service of doing "snapshot restoration" where pages that we proxy from Rails are snapshotted in memory right before page navigation so that back button can be instantaneous and any local DOM changes are kept.

So being able to turn off Vike's `onpopstate` and have our own would be nice, but so far not necessary.

**Comment by brillout** [comment]
Ideally, the user should be able to directly call `history.pushState()` and Vike would "simply" get out of the way.

But it isn't as simple as it seems, for example:
 1. Initial page (rendered by Vike as usual).
 2. pushState user land (i.e. independent of Vike without Vike's popstate handling).
 3. pushState triggered by Vike (click on a link pointing to a Vike route, or `navigate()` call).

When the user navigates back from `3.` to `2.` then Vike should actually handle that (despite `2.` being triggered by `pushState` called directly by the user).

Let me think.

**Comment by brillout** [comment]
> Ideally, the user should be able to directly call `history.pushState()` and Vike would "simply" get out of the way.

A drawback is that I don't think it's possible without [monkey patching `history.pushState()`](https://stackoverflow.com/questions/4570093/how-to-get-notified-about-changes-of-the-history-via-history-pushstate/4585031#4585031).

**Comment by dajinchu** [comment]
> When the user navigates back from 3. to 2. then Vike should actually handle that (despite 2. being triggered by pushState called directly by the user).

And Vike would handle it, right? since it listens to popstate? Is the issue that the history stack would be missing state that vike relies on?

**Comment by brillout** [comment]
AFAICT, regardless of what solution we go for, Vike always needs to listen to `popstate`. Not only that but it also needs to know everything about the history state. I don't think there is any alternative here. (The example of my previous comment is a good representation of why Vike needs to know everything.)

A call `navigate(url, { data: false, render: false })` would be equivalent to the user directly calling `history.pushState()`.

If Vike supports `navigate(url, { data: false, render: false })` then Vike might as well support the user directly calling `history.pushState()`. (I actually prefer telling users to directly call `history.pushState()` instead of adding a second option `render: boolean`, because it makes it clearer that the user has to implement its own `popstate` handler.)

A drawback is that it requires monkey patching `history.pushState()`. The alternative to monkey patching is to tell users to never use `history.pushState` which is, well, a net loss. So we might as well monkey patch it.

The good news is that I thought through all edge cases and, so far, it seems feasible.

The bad news is that it isn't trivial and I don't think there is a quick/dirty/trivial implementation that wouldn't break the browser's back-/forward navigation button.

> It seems like adding a shallow option to `navigate` would be simpler?

It doesn't really reduce complexity: we'd still have the issue showcased by the example of my previous comment (that's the bulk of the complexity).

**Comment by brillout** [comment]
I'll give the implementation a try and report back how it goes. ETA this week (but feel free to let me know if it's more urgent).

**Comment by brillout** [comment]
I'll give the implementation a try and report back how it goes. ETA this week. Does that work for you? Feel free to let me know if it's more urgent.

**Comment by dajinchu** [comment]
Thanks, really appreciate the support! If its any easier, we really only need `navigate(url, { data: false })` - having Vike continue to call onClientRender is fine.

**Comment by thomasjm** [comment]
Glad to see progress being made on this!

> A user tried before but [he ran into issues.](https://github.com/vikejs/vike/issues/1169#issuecomment-1761331578)

I actually found a solution that works well. I just commented on the other thread about it.

**Comment by brillout** [comment]
@thomasjm Thanks for circling back on this 👍 

@dajinchu I'm still on it. The internal refactor needed for this was quite extensive. That said, the refactoring is now done and I should now be able to experiment the simple implementation I've mind in order to enable users to directly use `history.pushState()` (it requires monkey patching it; I don't think any alternative is possible).

**Comment by brillout** [comment]
Just released `0.4.144-commit-7f5e99a`: you should now be able to use `history.pushState()`.

@dajinchu Let me know if it works for you and I'll solidify that functionality with some unit tests.

**Comment by dajinchu** [comment]
Thanks @brillout! We're testing it out now

**Comment by dajinchu** [comment]
@brillout Apologies for the delay on giving feedback. Seems this change introduced a regression on back/forward browser buttons, which are being counted as "isUserLandPushStateNavigation" and so if you use them on a parameterized route it fails to rerender.

Reproduction:
1. run `examples/react-full-v1`
2. http://localhost:3000/hello/eli
3. Click `/hello/jon` link
4. Click browser back button
5. Page still says "hello Eli"

**Comment by brillout** [comment]
Indeed I could reproduce, let me have a look.

**Comment by brillout** [comment]
Alright, this is now documented, unit tested, and battle tested by @dajinchu.

I think we can close this as completed: AFAICT allowing users to directly call `history.pushState()` is a superior design to Next.js's `shallow` feature (feel free to disagree).

Feel free to comment here if `history.pushState()` doesn't work out for you.

As always: [sponsoring](https://github.com/sponsors/vikejs) welcome.
